src/nnet3/nnet-simple-component.cc

```diff
@@ -186,7 +187,111 @@ const BaseFloat NormalizeComponent::kNormFloor = pow(2.0, -66);
+
+void NormalizeComponent::InitFromConfig(ConfigLine *cfl) {
+  int32 dim = 0;
+  BaseFloat target_rms = 1.0;
+  bool ok = cfl->GetValue("dim", &dim);
+  cfl->GetValue("target-rms", &target_rms);
+  if (!ok || cfl->HasUnusedValues() || dim <= 0) 
+    KALDI_ERR << "Invalid initializer for layer of type "           
+              << Type() << ": \"" << cfl->WholeLine() << "\"";
+  Init(dim, target_rms);
+}
+void NormalizeComponent::Read(std::istream &is, bool binary) {
+  std::ostringstream ostr_beg, ostr_end;
+  ostr_beg << "<" << Type() << ">"; 
+  ostr_end << "</" << Type() << ">"; 
+  ExpectOneOrTwoTokens(is, binary, ostr_beg.str(), "<Dim>");
+  ReadBasicType(is, binary, &dim_); // Read dimension.
+  std::string tok; // TODO: remove back-compatibility code.
+  ReadToken(is, binary, &tok);
+
+  // read target_rms_ if it is available.
+  if (tok == "<TargetRms>") {
+    ReadBasicType(is, binary, &target_rms_);
+    ReadToken(is, binary, &tok);
+  }
+
+  if (tok == "<ValueSum>") {
+    // this branch is for back compatibility.  TODO: delete it
+    // after Dec 2015.
+    value_sum_.Read(is, binary);
+    ExpectToken(is, binary, "<DerivSum>");
+    deriv_sum_.Read(is, binary);
+    ExpectToken(is, binary, "<Count>");
+    ReadBasicType(is, binary, &count_);
+    ExpectToken(is, binary, ostr_end.str());
+  } else {
+    // The new format is more readable as we write values that are normalized by
+    // the count.
+    KALDI_ASSERT(tok == "<ValueAvg>");
+    value_sum_.Read(is, binary);
+    ExpectToken(is, binary, "<DerivAvg>");
+    deriv_sum_.Read(is, binary);
+    ExpectToken(is, binary, "<Count>");
+    ReadBasicType(is, binary, &count_);
+    value_sum_.Scale(count_);
+    deriv_sum_.Scale(count_);
+    ExpectToken(is, binary, ostr_end.str());
+  }
+}
 
+void NormalizeComponent::Write(std::ostream &os, bool binary) const {
+  std::ostringstream ostr_beg, ostr_end;
+  ostr_beg << "<" << Type() << ">";
+  ostr_end << "</" << Type() << ">"; 
+  WriteToken(os, binary, ostr_beg.str());
+  WriteToken(os, binary, "<Dim>");
+  WriteBasicType(os, binary, dim_);
+  WriteToken(os, binary, "<TargetRms>");
+  WriteBasicType(os, binary, target_rms_); 
+  // Write the values and derivatives in a count-normalized way, for
+  // greater readability in text form.
+  WriteToken(os, binary, "<ValueAvg>");
+  Vector<BaseFloat> temp(value_sum_);
+  if (count_ != 0.0) temp.Scale(1.0 / count_);
+  temp.Write(os, binary);
+  WriteToken(os, binary, "<DerivAvg>");
+
+  temp.Resize(deriv_sum_.Dim(), kUndefined);
+  temp.CopyFromVec(deriv_sum_);
+  if (count_ != 0.0) temp.Scale(1.0 / count_);
+  temp.Write(os, binary);
+  WriteToken(os, binary, "<Count>");
+  WriteBasicType(os, binary, count_);
+  WriteToken(os, binary, ostr_end.str());
+}
+
+std::string NormalizeComponent::Info() const {
+  std::stringstream stream;
+  KALDI_ASSERT(InputDim() == OutputDim());  // always the case
+  stream << Type() << ", dim=" << dim_ 
+         << ", target_rms=" << target_rms_;
+
```

Comment:
If most of this is just a duplication of NonlinearComponent::Info(), why don't you do
stream << NonlinearComponent::Info();
to get the common parts, and avoid code duplication?